client/src/components/Tool/ToolForm.vue

```diff
@@ -43,6 +43,7 @@
                                 <FormDisplay
                                     :id="toolId"
                                     :inputs="formConfig.inputs"
+                                    :errors="formConfig.warnings || {}"
```

Comment:
I think you need to include errors as well:
<img width="839" alt="Screenshot 2023-05-18 at 16 42 52" src="https://github.com/galaxyproject/galaxy/assets/6804901/59a271c4-9732-4368-a0a7-1a14a73535e5">

that doesn't fix the wrong dataset selection, but this is probably also required. Given the complex interaction between backend and frontend an integration test is probably a good idea here.

Reply:
I added a test to verify that this works.

Reply:
Also added the initial errors to the form initialization.